### PR TITLE
Fix goroutine and websocket leak in logsink handler

### PR DESCRIPTION
**Pull request by babbageclunk**
## Description of change

If the ServeHTTP handler finishes before the receiveLogs goroutine, it can be stuck in the select trying to send on logCh. Add a stop channel to the goroutine that is always closed when the handler finishes and include that in the select.

This should fix the leak we see in prodstack, although we don't understand why it's happening now - none of the logsink code has changed between 2.4 and 2.5

## QA steps

* Basic smoketest - bootstrapped a controller, deployed some applications and ensured that logging was still getting to the database using debug-log.
* Haven't verified the goroutine leak because we haven't been able to reproduce that.

## Documentation changes
None

## Bug reference
Fixes https://bugs.launchpad.net/juju/+bug/1813104

### CODE REVIEW

**Comment by wallyworld** [comment]
Needs a test before landing

**Comment by babbageclunk** [comment]
$$merge$$

File: apiserver/logsink/logsink.go [hunk 1]
```diff
@@ -156,6 +167,9 @@ func (h *logSinkHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 				}
 			case m, ok := <-logCh:
 				if !ok {
+					h.mu.Lock()
+					defer h.mu.Unlock()
```
Comment (jameinel) [inline apiserver/logsink/logsink.go]: I think this is clearer as
mu.Lock()
var = true
mu.Unlock()

no need for defer when you are done right away.

File: apiserver/logsink/logsink.go [hunk 2]
```diff
@@ -233,6 +247,10 @@ func (h *logSinkHandler) receiveLogs(socket *websocket.Conn, endpointVersion int
 			// Send the log message.
 			select {
 			case <-h.abort:
+				// The API server is stopping.
+				return
+			case <-stop:
+				// The ServeHTTP handler has stopped.
 				return
 			case logCh <- m:
```
Comment (jameinel) [inline apiserver/logsink/logsink.go]: so the failure is that the api server isnt stopping but the log channel is?

**Comment by jameinel** [comment]
We ping once every 60s, so it is possible that is why I wasn't able to reproduce this. The failure seems to only be if the Ping to ensure the connection is alive fails.

**Pull request merged**